File: p49/interpreter/pkg/parser/parser.go
package parser

import (
	"fmt"
	"os"
	"regexp"
	"strings"

	"github.com/wuwenbo/fsm-dsl/interpreter/pkg/types"
)

type Token struct {
	Type   string
	Value  string
	Line   int
	Column int
}

type Parser struct {
	tokens []Token
	pos    int
}

func ParseDSL(input string) (*types.StateMachine, error) {
	tokens, err := tokenize(input)
	if err != nil {
		return nil, err
	}

	parser := &Parser{tokens: tokens, pos: 0}
	return parser.parse()
}

func tokenize(input string) ([]Token, error) {
	var tokens []Token
	line := 1
	column := 1

	keywords := map[string]bool{
		"machine": true, "state": true, "action": true,
		"initial": true, "on": true, "Enter": true, "Exit": true, "print": true,
	}

	re := regexp.MustCompile(`\s+|#[^\n]*|->|".*?"|[a-zA-Z_][a-zA-Z0-9_]*|[{}:(),]`)
	matches := re.FindAllStringIndex(input, -1)

	lastPos := 0
	for _, match := range matches {
		start, end := match[0], match[1]

		for i := lastPos; i < start; i++ {
			if input[i] == '\n' {
				line++
				column = 1
			} else {
				column++
			}
		}

		tokenStr := input[start:end]
		tokenColumn := column

		for i, c := range tokenStr {
			if c == '\n' {
				line++
				column = 1
			} else {
				column++
			}
			if i == 0 {
				tokenColumn = column
			}
		}

		tokenStr = strings.TrimSpace(tokenStr)
		if tokenStr == "" || strings.HasPrefix(tokenStr, "#") {
			lastPos = end
			continue
		}

		tokenType := "IDENTIFIER"
		if keywords[tokenStr] {
			tokenType = "KEYWORD"
		} else if strings.HasPrefix(tokenStr, "\"") && strings.HasSuffix(tokenStr, "\"") {
			tokenType = "STRING"
			tokenStr = tokenStr[1 : len(tokenStr)-1]
		} else if tokenStr == "->" {
			tokenType = "ARROW"
		} else if len(tokenStr) == 1 {
			switch tokenStr[0] {
			case '{':
				tokenType = "LBRACE"
			case '}':
				tokenType = "RBRACE"
			case ':':
				tokenType = "COLON"
			case '(':
				tokenType = "LPAREN"
			case ')':
				tokenType = "RPAREN"
			}
		}

		tokens = append(tokens, Token{
			Type:   tokenType,
			Value:  tokenStr,
			Line:   line,
			Column: tokenColumn,
		})

		lastPos = end
	}

	tokens = append(tokens, Token{Type: "EOF", Line: line, Column: column})
	return tokens, nil
}

func (p *Parser) current() Token {
	return p.tokens[p.pos]
}

func (p *Parser) consume(expectedType string, expectedValue ...string) (Token, error) {
	token := p.current()
	if token.Type != expectedType {
		return token, fmt.Errorf("expected %s but got %s at line %d", expectedType, token.Type, token.Line)
	}
	if len(expectedValue) > 0 && token.Value != expectedValue[0] {
		return token, fmt.Errorf("expected %s but got %s at line %d", expectedValue[0], token.Value, token.Line)
	}
	p.pos++
	return token, nil
}

func (p *Parser) parse() (*types.StateMachine, error) {
	machine, err := p.parseMachine()
	if err != nil {
		return nil, err
	}

	for p.current().Type != "EOF" {
		if p.current().Type == "KEYWORD" && p.current().Value == "action" {
			action, err := p.parseAction()
			if err != nil {
				return nil, err
			}
			machine.Actions = append(machine.Actions, *action)
		} else {
			p.pos++
		}
	}

	return machine, nil
}

func (p *Parser) parseMachine() (*types.StateMachine, error) {
	if _, err := p.consume("KEYWORD", "machine"); err != nil {
		return nil, err
	}

	nameToken, err := p.consume("IDENTIFIER")
	if err != nil {
		return nil, err
	}

	if _, err := p.consume("LBRACE"); err != nil {
		return nil, err
	}

	if _, err := p.consume("KEYWORD", "initial"); err != nil {
		return nil, err
	}

	if _, err := p.consume("COLON"); err != nil {
		return nil, err
	}

	initialToken, err := p.consume("IDENTIFIER")
	if err != nil {
		return nil, err
	}

	var states []types.State
	for p.current().Type != "RBRACE" && p.current().Type != "EOF" {
		if p.current().Type == "KEYWORD" && p.current().Value == "state" {
			state, err := p.parseState()
			if err != nil {
				return nil, err
			}
			states = append(states, *state)
		} else {
			p.pos++
		}
	}

	if _, err := p.consume("RBRACE"); err != nil {
		return nil, err
	}

	return &types.StateMachine{
		Name:    nameToken.Value,
		Initial: initialToken.Value,
		States:  states,
		Actions: []types.Action{},
	}, nil
}

func (p *Parser) parseState() (*types.State, error) {
	if _, err := p.consume("KEYWORD", "state"); err != nil {
		return nil, err
	}

	nameToken, err := p.consume("IDENTIFIER")
	if err != nil {
		return nil, err
	}

	if _, err := p.consume("LBRACE"); err != nil {
		return nil, err
	}

	state := &types.State{
		Name:        nameToken.Value,
		Transitions: []types.Transition{},
	}

	for p.current().Type != "RBRACE" && p.current().Type != "EOF" {
		if p.current().Type == "KEYWORD" && p.current().Value == "on" {
			p.pos++
			nextToken := p.current()

			if nextToken.Type == "KEYWORD" && nextToken.Value == "Enter" {
				p.pos++
				if _, err := p.consume("COLON"); err != nil {
					return nil, err
				}
				actionToken, err := p.consume("IDENTIFIER")
				if err != nil {
					return nil, err
				}
				state.OnEnter = actionToken.Value
			} else if nextToken.Type == "KEYWORD" && nextToken.Value == "Exit" {
				p.pos++
				if _, err := p.consume("COLON"); err != nil {
					return nil, err
				}
				actionToken, err := p.consume("IDENTIFIER")
				if err != nil {
					return nil, err
				}
				state.OnExit = actionToken.Value
			} else if nextToken.Type == "IDENTIFIER" {
				event := nextToken.Value
				p.pos++

				if p.current().Type == "COLON" {
					p.pos++
					actionToken, err := p.consume("IDENTIFIER")
					if err != nil {
						return nil, err
					}
					state.Transitions = append(state.Transitions, types.Transition{
						Event:  event,
						Action: actionToken.Value,
					})
				} else if p.current().Type == "ARROW" {
					p.pos++
					targetToken, err := p.consume("IDENTIFIER")
					if err != nil {
						return nil, err
					}
					state.Transitions = append(state.Transitions, types.Transition{
						Event:  event,
						Target: targetToken.Value,
					})
				}
			}
		} else {
			p.pos++
		}
	}

	if _, err := p.consume("RBRACE"); err != nil {
		return nil, err
	}

	return state, nil
}

func (p *Parser) parseAction() (*types.Action, error) {
	if _, err := p.consume("KEYWORD", "action"); err != nil {
		return nil, err
	}

	nameToken, err := p.consume("IDENTIFIER")
	if err != nil {
		return nil, err
	}

	if _, err := p.consume("LBRACE"); err != nil {
		return nil, err
	}

	var code strings.Builder
	braceCount := 1

	for braceCount > 0 && p.current().Type != "EOF" {
		if p.current().Type == "LBRACE" {
			braceCount++
		}
		if p.current().Type == "RBRACE" {
			braceCount--
		}
		if braceCount > 0 {
			code.WriteString(p.current().Value)
			code.WriteString(" ")
		}
		p.pos++
	}

	return &types.Action{
		Name: nameToken.Value,
		Code: strings.TrimSpace(code.String()),
	}, nil
}

func ParseFile(filename string) (*types.StateMachine, error) {
	content, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	return ParseDSL(string(content))
}

File: p49/language-server/src/parser.ts
import { StateMachine, State, Transition, Action } from '../../shared/types';

export class Token {
  type: string;
  value: string;
  line: number;
  column: number;

  constructor(type: string, value: string, line: number, column: number) {
    this.type = type;
    this.value = value;
    this.line = line;
    this.column = column;
  }
}

export class Lexer {
  private input: string;
  private pos: number = 0;
  private line: number = 1;
  private column: number = 1;

  private keywords = ['machine', 'state', 'action', 'initial', 'on', 'Enter', 'Exit', 'print'];

  constructor(input: string) {
    this.input = input;
  }

  tokenize(): Token[] {
    const tokens: Token[] = [];
    while (this.pos < this.input.length) {
      const char = this.input[this.pos];
      
      if (/\s/.test(char)) {
        if (char === '\n') {
          this.line++;
          this.column = 1;
        } else {
          this.column++;
        }
        this.pos++;
        continue;
      }

      if (char === '#') {
        while (this.pos < this.input.length && this.input[this.pos] !== '\n') {
          this.pos++;
        }
        continue;
      }

      if (char === '"' || char === "'") {
        tokens.push(this.readString(char));
        continue;
      }

      if (/[a-zA-Z_]/.test(char)) {
        tokens.push(this.readIdentifier());
        continue;
      }

      if (/[0-9]/.test(char)) {
        tokens.push(this.readNumber());
        continue;
      }

      const symbolTokens = this.readSymbol();
      if (symbolTokens) {
        tokens.push(...symbolTokens);
        continue;
      }

      this.pos++;
      this.column++;
    }
    tokens.push(new Token('EOF', '', this.line, this.column));
    return tokens;
  }

  private readIdentifier(): Token {
    const start = this.pos;
    const startColumn = this.column;
    while (this.pos < this.input.length && /[a-zA-Z0-9_]/.test(this.input[this.pos])) {
      this.pos++;
      this.column++;
    }
    const value = this.input.slice(start, this.pos);
    const type = this.keywords.includes(value) ? 'KEYWORD' : 'IDENTIFIER';
    return new Token(type, value, this.line, startColumn);
  }

  private readString(quote: string): Token {
    const startColumn = this.column;
    this.pos++;
    this.column++;
    const start = this.pos;
    while (this.pos < this.input.length && this.input[this.pos] !== quote) {
      if (this.input[this.pos] === '\n') {
        this.line++;
        this.column = 0;
      }
      this.pos++;
      this.column++;
    }
    const value = this.input.slice(start, this.pos);
    this.pos++;
    this.column++;
    return new Token('STRING', value, this.line, startColumn);
  }

  private readNumber(): Token {
    const start = this.pos;
    const startColumn = this.column;
    while (this.pos < this.input.length && /[0-9.]/.test(this.input[this.pos])) {
      this.pos++;
      this.column++;
    }
    return new Token('NUMBER', this.input.slice(start, this.pos), this.line, startColumn);
  }

  private readSymbol(): Token[] | null {
    const char = this.input[this.pos];
    const startColumn = this.column;
    const twoChar = this.input.slice(this.pos, this.pos + 2);
    
    if (twoChar === '->') {
      this.pos += 2;
      this.column += 2;
      return [new Token('ARROW', '->', this.line, startColumn)];
    }

    const symbols: { [key: string]: string } = {
      '{': 'LBRACE',
      '}': 'RBRACE',
      ':': 'COLON',
      ',': 'COMMA',
      '(': 'LPAREN',
      ')': 'RPAREN',
      '.': 'DOT'
    };

    if (symbols[char]) {
      this.pos++;
      this.column++;
      return [new Token(symbols[char], char, this.line, startColumn)];
    }

    return null;
  }
}

export class Parser {
  private tokens: Token[];
  private pos: number = 0;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  private current(): Token {
    return this.tokens[this.pos];
  }

  private consume(type: string, value?: string): Token {
    const token = this.current();
    if (token.type !== type || (value && token.value !== value)) {
      throw new Error(`Expected ${type}${value ? ` ${value}` : ''} but got ${token.type} ${token.value} at line ${token.line}`);
    }
    this.pos++;
    return token;
  }

  private peek(): Token {
    return this.tokens[this.pos + 1];
  }

  parse(): StateMachine {
    const machine = this.parseMachine();
    const actions: Action[] = [];
    
    while (this.current().type !== 'EOF') {
      if (this.current().type === 'KEYWORD' && this.current().value === 'action') {
        actions.push(this.parseAction());
      } else {
        this.pos++;
      }
    }

    return { ...machine, actions };
  }

  private parseMachine(): StateMachine {
    this.consume('KEYWORD', 'machine');
    const name = this.consume('IDENTIFIER').value;
    this.consume('LBRACE');
    
    this.consume('KEYWORD', 'initial');
    this.consume('COLON');
    const initial = this.consume('IDENTIFIER').value;

    const states: State[] = [];
    while (this.current().type !== 'RBRACE' && this.current().type !== 'EOF') {
      if (this.current().type === 'KEYWORD' && this.current().value === 'state') {
        states.push(this.parseState());
      } else {
        this.pos++;
      }
    }
    this.consume('RBRACE');

    return { name, initial, states, actions: [] };
  }

  private parseState(): State {
    this.consume('KEYWORD', 'state');
    const name = this.consume('IDENTIFIER').value;
    this.consume('LBRACE');

    const state: State = {
      name,
      transitions: []
    };

    while (this.current().type !== 'RBRACE' && this.current().type !== 'EOF') {
      if (this.current().type === 'KEYWORD' && this.current().value === 'on') {
        this.pos++;
        const nextToken = this.current();
        
        if (nextToken.type === 'KEYWORD' && nextToken.value === 'Enter') {
          this.pos++;
          this.consume('COLON');
          state.onEnter = this.consume('IDENTIFIER').value;
        } else if (nextToken.type === 'KEYWORD' && nextToken.value === 'Exit') {
          this.pos++;
          this.consume('COLON');
          state.onExit = this.consume('IDENTIFIER').value;
        } else if (nextToken.type === 'IDENTIFIER') {
          const event = nextToken.value;
          this.pos++;
          
          if (this.current().type === 'COLON') {
            this.pos++;
            const action = this.consume('IDENTIFIER').value;
            state.transitions.push({ event, target: '', action });
          } else if (this.current().type === 'ARROW') {
            this.pos++;
            const target = this.consume('IDENTIFIER').value;
            state.transitions.push({ event, target });
          }
        }
      } else {
        this.pos++;
      }
    }
    this.consume('RBRACE');

    return state;
  }

  private parseAction(): Action {
    this.consume('KEYWORD', 'action');
    const name = this.consume('IDENTIFIER').value;
    this.consume('LBRACE');
    
    let code = '';
    let braceCount = 1;
    while (braceCount > 0 && this.current().type !== 'EOF') {
      if (this.current().type === 'LBRACE') braceCount++;
      if (this.current().type === 'RBRACE') braceCount--;
      if (braceCount > 0) {
        code += this.current().value + ' ';
      }
      this.pos++;
    }

    return { name, code: code.trim() };
  }
}

export function parseDSL(input: string): { ast: StateMachine; tokens: Token[] } {
  const lexer = new Lexer(input);
  const tokens = lexer.tokenize();
  const parser = new Parser(tokens);
  const ast = parser.parse();
  return { ast, tokens };
}
